File: ios/sail-ios/JsonRpcWatchPluginContext.m
//
//  JsonRpcWatchPluginContext.m
//  sail-ios
//
//  Copyright (c) 2012 KTH Corp.
//

#import "JsonRpcWatchPluginContext.h"
#import "AxLog.h"
#import "JSONKit.h"
#import "BridgeSessionManager.h"

#define WATCH_SAMPLE_METHOD     @"ax.watch.sample"

@implementation JsonRpcWatchPluginContext

- (id)initWithRequestJson:(NSDictionary*)json session:(BridgeSession*)session {
    return self = [super initWithRequestJson:json session:session];
}

#pragma mark -

- (void)sendResult:(id)result {
    AX_LOG_TRACE(@"sendResult(%@) called on JsonRpcWatchPluginContext(%d). it will be discarded.", [result description], [self.identifier intValue]);
}

- (void)sendError:(NSInteger)code message:(NSString *)message {
    [self sendWatchError:code message:message];
}

- (void)sendRpcNotify:(NSString*)method withParam:(id)param {
    NSDictionary* map = [NSDictionary dictionaryWithObjectsAndKeys:
                         [NSNull null], @"id",
                         method, @"method",
                         [NSArray arrayWithObjects:param, nil], @"params",
                         nil];
    [super sendRpcResponse:map];
}

- (void)sendWatchResult:(id)result {
    [super makeSuccessResult:result];
    [self sendRpcNotify:WATCH_SAMPLE_METHOD withParam:self.result];
}

- (void)sendWatchError:(NSInteger)code message:(NSString *)message {
    [super makeErrorResult:code message:message];
    [self sendRpcNotify:WATCH_SAMPLE_METHOD withParam:self.result];
}

@end

File: android/sail-android/src/com/appspresso/core/runtime/server/kraken/JsonRpcWatchPluginContext.java
package com.appspresso.core.runtime.server.kraken;

import java.util.HashMap;
import java.util.Map;

import org.apache.commons.logging.Log;

import com.appspresso.api.AxLog;
import com.appspresso.core.runtime.server.kraken.BridgeSessionManager.BridgeSession;
import com.appspresso.core.runtime.util.JsonUtil;

public class JsonRpcWatchPluginContext extends JsonRpcAsyncPluginContext {

    private final static Log L = AxLog.getLog("JsonRpcWatchPluginContext");

    private static final String WATCH_SAMPLE_METHOD = "ax.watch.sample";

    public JsonRpcWatchPluginContext(Map<?, ?> requestJson, BridgeSession session) {
        super(requestJson, session);
    }

    @Override
    public void sendResult(Object object) {
        if (L.isTraceEnabled()) {
            L.trace(String
                    .format("sendResult(%s) called on JsonRpcWatchPluginContext(%d). it will be discarded.",
                            JsonUtil.toJson(object), getId()));
        }
    }

    @Override
    public void sendError(int code, String message) {
        sendWatchError(code, message);
    }

    @Override
    public void sendWatchResult(Object object) {
        super.makeSuccessResult(object);
        sendRpcNotify(WATCH_SAMPLE_METHOD, super.getResult());
    }

    @Override
    public void sendWatchError(int code, String message) {
        super.makeErrorResult(code, message);
        sendRpcNotify(WATCH_SAMPLE_METHOD, super.getResult());
    }

    private void sendRpcNotify(String method, Object... params) {
        Map<String, Object> map = new HashMap<String, Object>();
        map.put("id", null);
        map.put("method", method);
        map.put("params", params);
        sendRpcResponse(map);
    }

}
